cluster/evaluator/evaluator_ringmaster_wrapper.py

```diff
@@ -0,0 +1,215 @@
+import bigtable_output
+from bigtable_input import METADATA
+
+
+CTL_NAME = "ringmaster_evals"
+CTL_FILENAME = CTL_NAME + ".ctl"
+CTL_GAME_DIR = CTL_NAME + ".games"
+CTL_LOG = CTL_NAME + ".log"
+CTL_REPORT = CTL_NAME + ".report"
+
+MODEL_ROW_FMT = "m_eval_{}"
+
+ENV_VARS = [
+    "MODEL_WHITE",
+    "MODEL_BLACK",
+    "SGF_BUCKET_NAME",
+    "PROJECT",
+    "CBT_INSTANCE",
+    "CBT_TABLE",
+]
+ENV = {}
+for env_var in ENV_VARS:
+    value = os.environ.get(env_var)
+    assert value, (env_var, os.environ.keys())
+    ENV[env_var] = value
+
+Model = namedtuple("Model", ["name", "og_model", "model_name", "flags"])
+
+def setup_ringmaster(model_a, model_b, start_time):
+    CTL_FILE = """
+competition_type = "playoff"
+description = " Testing models_for_eval "
+board_size = 19
+komi = 7.5
+
+record_games = True
+stderr_to_log = True
+
+def MinigoPlayer(path, model_pb, flags):
+    return Player(
+        "bin/bazel-bin/cc/gtp --model='tf,{{}}' {{}}".format(model_pb, flags),
+        cwd=path,
+        environ={{"LD_LIBRARY_PATH": "bin/cc/tensorflow/lib"}},
+        sgf_player_name_from_gtp=False)
+
+players = {{
+    "{m_a.name}": MinigoPlayer("{m_a.name}", "{m_a.model_name}", "{m_a.flags}"),
+    "{m_b.name}": MinigoPlayer("{m_b.name}", "{m_b.model_name}", "{m_b.flags}"),
+}}
+
+matchups = [
+    Matchup("{m_a.name}", "{m_b.name}",
+            id="{m_a.name}_vs_{m_b.name}_t_{start_time}",
+            alternating=True,
+            number_of_games={num_games})
+]
+"""
+
+    with open(CTL_FILENAME, "w") as ctl_f:
+        ctl_f.write(CTL_FILE.format(
+            m_a = model_a,
+            m_b = model_b,
+            start_time = start_time,
+            num_games = 4))
+
+
+def call_ringmaster():
+    process = subprocess.run(
+        ["ringmaster", CTL_FILENAME, "run"],
+        capture_output=False,
+        stderr=subprocess.STDOUT,
+        timeout=20*60)
+    # log errors somewhere
+    return process.returncode == 0
+
+
+def copy_to_gcs(src, dst):
+    assert gfile.Exists(src)
+    assert not gfile.Exists(dst)
+
+    with gfile.GFile(src, "rb") as src_f, gfile.GFile(dst, "wb") as dst_f:
+        shutil.copyfileobj(src_f, dst_f)
+
+
+def record_results(bt_table, sgf_base, num_games, start_time):
+    games = os.listdir(CTL_GAME_DIR)
+    failed = len(games) != num_games
+
+    if failed:
+        # Upload something? log error somewhere?
+        assert False, (len(games), num_games)
+
+    # Upload .log and .report along side all .games
+    copy_to_gcs(CTL_LOG, os.path.join(sgf_base, CTL_LOG))
+    copy_to_gcs(CTL_REPORT, os.path.join(sgf_base, CTL_REPORT))
+
+    rows = []
+    for game_fn in games:
+        game_path = os.path.join(CTL_GAME_DIR, game_fn)
+        copy_to_gcs(game_path, os.path.join(sgf_base, CTL_REPORT))
+
+        metadata = bigtable_output.process_game(game_path)
+        metadata["sgf"] = game_fn
+        metadata["tool"] = "evaluator_ringmaster"
+
+        # game_fn contains timestamp and is hence unique.
```

Comment:
games don't finish at the same second?

Reply:
game_filename will be "<model>_vs_<model>_<script start time>_<gamenum>.sgf"

Reply:
perfect